### ISSUE TITLE
reportTypedDictNotRequiredAccess false-positive: LBYL doesn't work when iterating over dict keys

### ISSUE THREAD

**Issue by jshumphrey** [microsoft/pyright#11206]
**Describe the bug**
When iterating over a tuple of `TypedDict` key values, Pyright emits `reportTypedDictNotRequiredAccess` when accessing a non-required `TypedDict` key even when LBYL-ing via `if field_name in typed_dict_instance` (as recommended by [the Pylance docs for `reportTypedDictNotRequiredAccess`](https://github.com/microsoft/pylance-release/blob/main/docs/diagnostics/reportTypedDictNotRequiredAccess.md)).

**Code or Screenshots**
Code sample in [pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBDAnmYA7A5gKEwYwDYCGAzkXALKIAqyApgCYAiwOMAwoSQBRIoYB01MPSYsAlAC5McaXCg0AjgFdgcuuLhEYUKTNQRUAfTlKV9AwEZ1PNOj4A5CDABKC5aoDamqAF0d0vYbGbmYATFbINvaOLiYeXr6YIIgGPGZ0zDAGaJoEqDg0cAC85FS0jBnsxEScAN5%2BcABEQaZ0DepNBgBuBHiKNA0ANPUAxHABRq4tFhoA1sBgQnQDcABGivAwABYE8HQQNESoAOTwRDQ0IAgQcCA7MDRQcLmIAO4EiPUN482qBiFtjVQUBCXR6fUGmAAvqJsMAAGaA-QTWJmcwNOBoG7JVJ0AzpFhZVA5PI0dRwUYASQRWwKMxoiAxpDwwHuUB6QxkcAMRSxKTKeIyhOJ%2BXcXyRP1RDW80lGNNQGJgR1IEBm72wcOgcDpDMxnCak1Ug0RgQNkuWYpNKNx-wk9XhWvpGPlST5iwFBOyMFy%2BTJowAQus4C9NjR5SyHjsbHAIJ0Hg7EEQOZyZFUHvBtU7eTj3ZlPd6CuS4JxU7A4BrHugIBA6DcaMRFHIYcmZNzii7s-jc0SvST3NrpYXEBBFHB0DR4Lk4A8oNBMEA)

```python
import typing

class MyTypedDictClass(typing.TypedDict):
    required: str
    non_required_1: typing.NotRequired[str]
    non_required_2: typing.NotRequired[str]

my_typed_dict_instance = MyTypedDictClass({
    "required": "r_value",
    # non_required_1 skipped, but that doesn't seem to matter anyway
    "non_required_2": "nr2_value",
})

if "non_required_1" in my_typed_dict_instance:  # If the key is literal,
    _ = my_typed_dict_instance["non_required_1"]  # then it's okay

for key in ("required", "non_required_1", "non_required_2"):
    if key in my_typed_dict_instance:  # But when iterating over keys,
        assert key in my_typed_dict_instance  # (assert for good measure)
        _ = my_typed_dict_instance[key]  # you get an error

```

**VS Code extension or command-line**
I'm running Pyright as part of the Pylance v2025.10.4 VS Code extension, but the issue is the same whether in VS Code or in the Pyright Playground.

**Comment by erictraut** [comment]
Yes, this behavior is expected. You're using a dynamic (variable) key to index the TypedDict. Pyright doesn't support type narrowing for dynamic keys. (Nor do any other Python type checkers.)

You can avoid the type error by using `_ = my_typed_dict_instance.get(key)`.
